Inference tests from master branch

notoraptor opened the pull request:
I reopend my PR here, as @breuleux PR is now merged.
- Still WIP, I will fix more tests today.
- Numpy dependencies removed.
- Basic support for dict added.
- currently 187 failed, 170 passed, 3 xfailed among added tests

notoraptor committed:
Get inference tests from master branch:
- tests/operations/test_operations
- tests/operations/test_universe
- tests/test_infer
Ignored:
- tests/composite/test_composite_examples (we don't have concept of composite y branch zero)
- tests/operations/test_macro_dtype (we don't yet have dtype op)
- tests/operations/test_macro_einsum (we don't yet have einsum op)

notoraptor committed:
Remove `run` tests from imported files.

notoraptor committed:
Add temporary placeholders to fullfill imports in test files.

notoraptor committed:
Fix errors in test_universe:
- remove unused and irrelevant imports
- remove unused _test_universe_chk
- remove now irrelevant test_universe_commit

notoraptor committed:
Add master placeholders for symbols from master branch but not (yet?) implemented in zero branch
Add more placeholders in myia/testing/common

Remove test_bad_metagraph (as no MetaGraph class is implemented)
Remove test_contextless (as no concept of contexless is implemented)
Remove _test_op (as concept of primitive is not implemented)
Remove test_bad_operation (as concept of operation is not implemented)
Remove pipleine-dependent code
Use ovld for `mystery` function

Replace x.head and x.tail with x[0] and x[-1] for abstract lists.

notoraptor committed:
Remove test_hastag_casttag (no tag concept implemented)

Remove test_J, test_Jinv (grad not yet implemented)

Remove unused abstract_inferrer_cons_test

Resolve placeholders, remove unused placeholders.

notoraptor committed:
Update tests. Bool type support arithmetic operations.

notoraptor committed:
Fix test_prim_tern.

notoraptor committed:
Fix test_prim_usub.

notoraptor committed:
Asd specific to_abstract() for module objects.
Make getattr_inferrer work correctly for module objects.
Make infer() tester support abstract union against abtract value in return type.

Add specific _unify() for an abstract union against an abstract value.

Partially fix test_prim_log (np.log):
- work for scalars
- does not work for arrays

notoraptor committed:
Remove all placeholders from common module.

Fix test_scalar_to_array.

notoraptor committed:
Separate failing test (for arrays) for test_prim_log

notoraptor committed:
Fix test_if2

notoraptor committed:
Fix test_not_enough_args_prim
Fix test_too_many_args_prim
Fix test_not_enough_args
Fix test_too_many_args

notoraptor committed:
Make inferrer support make_tuple, make_list, operator.getitem

Add support for more builtin operators

notoraptor committed:
Make inferrer detect functions that are types, then infer type as function return.

notoraptor committed:
Make inferer support global_universe_setitem
Make sur to infer type for unused nodes, by visiting SEQ nodes

Fix test_universe

notoraptor committed:
Correctly merge value track ANYTHING with other value.

Allow to add inferrers for classes and types.

Replace Tuple, List and Array with tuple_of(), list_of() and array_of() functions.

Use a function to add testing inferrers.

Add a "numpy_subset" module to handle used numpy symbols.

Fix test_infer_scalar_cast

notoraptor committed:
Fix test_infer_full.

notoraptor committed:
Fix test_if

notoraptor committed:
Partially fix test_prim_log

notoraptor committed:
Re-fix test_if
Fix test_prim_log

notoraptor committed:
Fix test_list

notoraptor committed:
Code cleanup and reformating.
Add Numpy to pyproject.

notoraptor committed:
Correctly display interface name in AbstractStructure.__hrepr__

notoraptor committed:
Improve printing of MapError.

notoraptor committed:
Remove numpy dependencies.

notoraptor committed:
Fix test_scalar_cast

notoraptor committed:
Move abstract_map2 to _unify for ValueTrack.

notoraptor committed:
Unify ValueTrack by returning ANYTHING if values are different.
Remove special handling for ValueTrack in abstract_map2 for AbstractAtom.

notoraptor committed:
Do not unify abstract union with a non-abstract union.

Use str and None instead of abstract atoms in inferrers definitions.

Make inference check result using a ovld-ed _check_inference(ret, expected) method.
- This should alow to handle unions vs atoms/structures in inference testing rather than unification.

Make switch inferrer return a type instead of a union if both branches returns same type.

Update test_prim_log to test_math_log.

notoraptor committed:
Add utilities to define multiple-signature inference.

Fix test_math_log

notoraptor committed:
Explicitly compare abstract atoms for return types in inference testing.

Fix test_if

notoraptor committed:
Move inference dispatcher to infnode.
Improve abstract atom comparison for inference testing result.

Partially fix test_while

notoraptor committed:
Improve _get_inference_function()

notoraptor committed:
Check bool value if available in user_switch inference.

Add inference for myia_iter, myia_hasnext and myia_next.

Improve InferenceDefinition printing and type parsing.

notoraptor committed:
Remove not implemented master operations:
- broadcast_shape
- embed
- env_getitem
- env_setitem
- gadd
- grad
- J
- Jinv
- make_record
- record_setitem
- tagged

notoraptor committed:
Add ovld-implementation for zeros_like
Separate zeros_like testing into passing and failing tests.

notoraptor committed:
Implement master placeholders:
- dict_values
- dict_setitem

Add a basic support for dicts.

A dict type is defined using sequence of flattened key-value types.

notoraptor committed:
Use InferenceDefinition.type_to_abstract to get abstract types in signature() function.

This allows to correctly convert `None` to `AbstractAtom({"interface": type(None)})`.

This fix universe tests.

notoraptor committed:
Fix test_len.

notoraptor committed:
test_tuple_getitem_negative fails because `-1` is not a constant.
Instead it's an apply node `operator.neg(1)`.

test_tuple_getslice fails because slice value is not a constant.
Instead it's an apply node `slice(...)`

test_concat_tuple fails because inference for `tuple + tuple` is not yet handled.

notoraptor committed:
Fix test_tuple_setitem

notoraptor committed:
Remove test_zeros_like_fn (no env logic in current zero branch code)

notoraptor committed:
Mark dataclasses test as xfail, as dataclasses are not yet supported.

notoraptor committed:
Add an inferrer for isinstance.

notoraptor committed:
Add inference for some missing builtin funcctions.
Sort inference functions.
Remove unused Shp() in test_operations.
Make sure to add all required parameters to test_arithmetic_data_truediv

notoraptor committed:
Pass inferrers to inference functions.

Implement a special generic inference for binary and unary operations, that will look for related Python magic methods in left operand class.

Fix test_concat_tuple
Fix test_func_arg3

Fix arithmetic tests on Point and Point3D classes (mark them as xfail)

notoraptor committed:
Add int.__neg__ inference
Will be called for bool.__neg__ inference too

Fix test_prim_usub

notoraptor committed:
Improve isinstance inferrer. Support unions.

notoraptor committed:
Fix test_call_nonfunc

notoraptor committed:
Fix test_unknown_method

notoraptor committed:
Mark test_getattr_union as xfail (dataclasses not yet supported)

notoraptor committed:
Make sur isinstance inference support tuples again

Fix test_isinstance_bad

notoraptor committed:
Fix test_tuple_outofbound_negative

notoraptor committed:
Add inference for float.__mul__

Fix test_constants_floatxint

notoraptor committed:
Fix test_unsafe_static_cast_error

notoraptor committed:
Remove unused newenv

(wip) Try to support getattr on classes (not objects) for static/class methods.

Some related tests:
- test_getattr
- test_getattr_multitype

notoraptor committed:
Apply flake8, pydocstyle, isort and black.

notoraptor commented:
Update.

For added tests:

`pytest -vvs tests/operations/ tests/test_infer.py`:
- `131 failed, 202 passed, 26 xfailed in 22.92s`

I think I won't improve it too much from now to meeting, so I stop here for now.

What do you think ? @breuleux

notoraptor committed:
Back to simple error printing so that error log can be read.

notoraptor committed:
Xfail test_default_arg

notoraptor committed:
Simplify binary and unary op generic inferences.

notoraptor committed:
Fallback to rop inferrers if lop inferrers don't exist for binary operations.

Allow InferenceDefinition to support default inferrer.

notoraptor committed:
Split test_list_and_scalar into passing and failing tests.

notoraptor committed:
Add basic support for constant slices in getitem inferrer for tuples.

Support lists in getitem inferrer for indices and slices.

Fix test_tuple_getslice

notoraptor committed:
Xfail test_closure

notoraptor committed:
Xfails test_return_closure

notoraptor committed:
Xfail test_varargs

notoraptor committed:
Xfail test_apply

notoraptor committed:
Xfail test_choose_prim

notoraptor committed:
Xfail test_choose_incompatible

notoraptor committed:
Xfail test_choose_indirect

notoraptor committed:
Xfail test_hof_2

notoraptor committed:
Xfail test_func_arg4

notoraptor committed:
Xfail test_isinstance

notoraptor committed:
Xfail test_union

notoraptor committed:
Xfail test_union_2

notoraptor committed:
Xfail test_union_nil
Xfail test_union_and

notoraptor committed:
Fix test_user_switch_hastype (raises InferenceError in master branch)

notoraptor committed:
Xfail test_partial_2

notoraptor committed:
Xfail test_recursive_build

notoraptor committed:
Xfail test_infinite_mutual_recursion

notoraptor committed:
Fix test_unknown_data

notoraptor committed:
Xfail test_union_binand

notoraptor committed:
Xfail test_closure_passing

Xfail test_closure_deep

notoraptor committed:
Xfail test_hof
Xfail test_hof_3

notoraptor committed:
Xfail test_keywords_bad_3 (parser error ?)
Xfail test_kwarg
Xfail test_defaults_recursive
Xfail test_redundant_kw
Xfail test_keywords_shadow
Xfail test_keywords_defaults
Xfail test_keywords_different_order
Xfail test_keywords_bad
Xfail test_keywords
Xfail test_keywords_expand
Xfail test_default_closure
Xfail test_multitype_function

notoraptor committed:
xfail:
- test_hastype_2
- test_choose_prim_incompatible
- test_hof_4
- test_hof_5

notoraptor committed:
Split test_hastype() into passing and failing tests.

notoraptor committed:
Xfail:
- test_union_nest
- test_union_nested_2
- test_hasattr_cond
- test_getattr
- test_getattr_multitype
- test_getattr_flex
- test_bool_and
- test_bool_or
- test_and
- test_and_none
- test_multitype
- test_nullary_closure
- test_merge_point
- test_dict_merge

Fix
- test_switch
- test_is
- test_is_not
- test_constants_floatxint2 (add inference for float.__radd__)

notoraptor committed:
Fix test_prim_tern.

notoraptor committed:
Fix tests/abstract/test_data::test_repr_tracks.

Depending on prior execution, this text may fail at line

`assert str(i.tracks) == "TrackDict(interface ↦ class int, value ↦ 1)"`

because tracks are not printed in same order (ie. "value" printed before "interface")

Fix it by sorting tracks when generating Tracks str.

notoraptor committed:
Update doc for _unify(ValueTrack, ValueTrack)
Remove _merge override for (ValueTrack, ValueTrack)

notoraptor committed:
Remove currently unused tracks ndim and shape.

notoraptor committed:
Add details about to_abstract(str)
Reformat code

notoraptor committed:
Fix tests.abstract.test_utils.test_cannot_unify:
Transfer passing tests to test_unify

Some tests pass now because ValueTrack are unified with ANYTHING when values are different.

notoraptor committed:
Fix tests.infer.test_algo.test_merge

test_merge now pass because value tracks can now be unified with ANYTHING

notoraptor committed:
Re-fix:
- test_merge
- test_unify
- test_generic
- test_sum
- test_sum_method
- test_format_exc
- test_format_exc_no_debug

notoraptor committed:
Simplify inferrers code.

notoraptor committed:
Remove useless `inferrers` parameters from inference functions.

notoraptor committed:
Fix tests.infer.test_infer.test_fact: Make bin op inference support generics.

notoraptor committed:
Fix coverage:
- No need to compare type when unifying ValueTrack.
- myia_hasnext and myia_next always receive expected types, already checked by myia_iter
- do not check that _bin_rop_inferrer receive different operand types
- InferenceDefinition.type_to_abtract never receive lists/tuples currently.
- remove unused symbols from common
- move dataclasses from common to test_infer (as they are used only there)
- Use A instead of Ex and External
- Inline scalar_cast and unsafe_static_cast
- Remove unused _check_inference() overload in multitest
- Test than we cannot unify dicts with different sizes.
- Add inference test for range
- Add inference test for list getitem (index and slice)
- Remove useless test_multitype

notoraptor committed:
Move master-imported codes into a single test folder.
Make sure to reinitialize truediv inferrer after running tests.infer.test_infer.test_specialization.
Reformat code.

notoraptor committed:
Make sure module symbol got from getattr is a function.
Check that abstract dict contains unique abstract keys.
Add more comment details in infnode.
Remove unused case in _get_inference_function.
Remove useless _to_abstract overload from common.
Remove useless _check_inference overload from multitest.
Revert confest changes
Reformat code.

notoraptor commented:
Well, I guess it's ready for a review and possible merging now, @breuleux !

breuleux commented on myia/abstract/data.py:
```diff
@@ -273,6 +279,33 @@ def __hrepr__(self, H, hrepr):
+    @property
+    def values(self):
+        """Get list of value types."""
+        return self.elements[1::2]
```
I would rather avoid having a specific `AbstractDict` type here. The basic idea behind Abstract is that all the subtypes are the elements, and the interface contains the necessary information to figure out what to do with them. In this case, for e.g. `{"a": 1, "b": 2.3}` I would use something like `AbstractStructure((int, float), {"interface": dict_with_keys("a", "b")})`. The int and float are the things we can't throw away because they are the types of the actual things we need to put in the dictionary, whereas the keys "a" and "b" we can eventually throw away along with the interface when we replace the dict by a simple tuple.

notoraptor replied:
OK ! I removed AbstractDict and just use an AbstractStructure instead. I keep the same code logic (ie. interface as `dict`, and elements as list of flattened key-value types, corresponding to what `make_dict` would receive).

This allows a basic support for dicts, but I guess dict support will be modified/updated later for more complex cases.

breuleux commented on myia/infer/inferrers.py:
```diff
@@ -47,52 +63,486 @@ def partial_inferrer(node, args, unif):
     return res
 
 
+def getitem_inferrer(node, args, unif):
```
Instead of special casing various interfaces, this can dispatch to `x.__getitem__(y)` through `Replace`, and the various builtin types' getitem methods can have inferrers.

notoraptor replied:
Done!

breuleux commented on myia/infer/inferrers.py:
```diff
@@ -47,52 +63,486 @@ def partial_inferrer(node, args, unif):
+    return data.AbstractDict(arg_types)
+
+
+def len_inferrer(node, args, unif):
```
This can just dispatch to `x.__len__()`.

notoraptor replied:
Shouldn't `x.__len__()` return an integer in any case ? If so, I think we could just check that interface has the `__len__` method, then assume return type is a `int`.

breuleux replied:
It is true that `len` forces its result to be an integer, but if the user defines their own `__len__` method and it returns something else than an integer, we would like to be able to report an error.

notoraptor replied:
OK, done !

breuleux commented on myia/infer/inferrers.py:
```diff
@@ -47,52 +63,486 @@ def partial_inferrer(node, args, unif):
+        return res
+
+
+def operator_add_inferrer(node, args, unif):
+    """Inferrer for the operator.add function."""
+    return _bin_op_inferrer("__add__", "__radd__", node, args, unif)
```
It would be more economical to have `_bin_op_inferrer("__add__", "__radd__")` return the inferrer, and write that directly in the inferrers directly. (Same for below).

notoraptor replied:
Done!

breuleux commented on tests/infer/from_master/master_inferrers.py:
```diff
@@ -0,0 +1,87 @@
+"""Supplementary inferrers for testing.
```
Could you keep `from_master` in a different branch? I don't see a purpose in merging these in.

notoraptor replied:
Done!

notoraptor committed:
Remove `from_master` inferrers and placeholders.

Add inference for type constructors `object` and None type.

Change test_math_log to test_module_function_call to keep testing inference for a module-imported function.

notoraptor committed:
Simplify inferrers for unary and binary operations.
- Write generic generators for inferrers functions and InferenceFunction objects.

notoraptor committed:
Rebase and fix tests.
- There are now more MyiaSyntaxErrors in inference tests.

notoraptor committed:
Replace getitem operator with call to left operand's __getitem__ method.

notoraptor committed:
Replace len function with a call to operand's __len__ method.

notoraptor committed:
Remove data.AbstractDict class. Simply use data.AbstractStructure.

notoraptor commented:
@breuleux PR updated!

breuleux commented on myia/abstract/map.py:
```diff
@@ -366,6 +369,12 @@ def abstract_map2(  # noqa: F811
     if len(x.elements) != len(y.elements):
         raise MapError(x, y, reason="Structures have different lengths")
 
+    if x.tracks.interface is dict:
+        # Just check keys, tracks will be checked below.
+        for kx, ky in zip(x.elements[::2], y.elements[::2]):
+            if kx is not ky:
+                raise MapError(kx, ky, "Different dict keys")
+
```
The keys should be in the interface, though (e.g. a class `DictWithKeys("x", "y")`), and if the keys are different, the interfaces would be different, so there would be no need to special case anything.

breuleux commented on myia/abstract/to_abstract.py:
```diff
@@ -20,11 +21,35 @@ def to_abstract(self, xs: tuple):
+    return data.AbstractAtom({"interface": type(x), "value": x})
+
+
+@ovld
+def to_abstract(self, x: dict):  # noqa: F811
+    return data.AbstractStructure(
+        [self(el) for item in x.items() for el in item], {"interface": dict}
+    )
```
They keys should not be in the structure's elements.

notoraptor committed:
Add a new class DictWithKeys to use as interface for abstract dictionaries.

notoraptor committed:
Rebase and fix.

notoraptor commented:
@breuleux PR updated about dicts, anyway !

I'm working on splitting it.
